### PR TITLE
New queries and mutations for verifying tokens and manipulating sessions

### PR DESCRIPTION
**Pull request by pozylon**
Depends on:
https://github.com/orbiting/republik-backend/pull/49

### CODE REVIEW

File: packages/auth/lib/signInWithToken.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      passport: {
+        user: user.id
+      }
+    }
+  })
+
+  // call signIn hooks
+  await Promise.all(
+    signInHooks.map(hook =>
+      hook(user.id, !existingUser, pgdb)
```
Comment (pozylon) [inline packages/auth/lib/signInWithToken.js]: @patte Your change is here, if you wonder

File: packages/auth/graphql/schema.js [hunk 2]
```diff
@@ -18,11 +18,16 @@ type queries {
   # search for users
   # required role: editor
   users(search: String!, role: String!): [User]!
+
+  # search for an unverified session by token
+  unverifiedSession(email: String!, token: String!, context: String): Session
```
Comment (pozylon) [inline packages/auth/graphql/schema.js]: I really hate the query name, what do you guys think? @patte @tpreusse 

- getInfo is far too unspecific to get session details
- session is not true as it only returns sessions that have a token and no passport.user set

More ideas: wildSession, sessionWithToken, unclaimedSession

Comment (tpreusse) [reply packages/auth/graphql/schema.js]: IMM: `authorizeSession` for the mutation and `unauthorizedSession` for the query

File: packages/auth/graphql/schema.js [hunk 3]
```diff
@@ -18,11 +18,16 @@ type queries {
...
 }
 
 type mutations {
   # signIn with an email address
   signIn(email: String!, context: String): SignInResponse!
   signOut: Boolean!
+  # validate a token sent by mail to convert a login request token to a valid user session
+  signInWithToken(email: String!, token: String!, context: String): Boolean!
```
Comment (tpreusse) [inline packages/auth/graphql/schema.js]: context not need here IMO

File: packages/auth/lib/hashSessionId.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module.exports = (id) => {
+  const crypto = require('crypto')
+  return crypto
+    .createHmac('sha256', process.env.SESSION_SECRET)
```
Comment (patte) [inline packages/auth/lib/hashSessionId.js]: don't use the SESSION_SECRET here. I could
- get my session ID from the dev tools
- get my HMACed session ID from the API
- start a bruteforce attach to get SESSION_SECRET
invent a independent ENV like SESSION_ID_HMAC and use checkENV to make sure it's set.

Comment (tpreusse) [reply packages/auth/lib/hashSessionId.js]: @patte random thought: create a secret per user and save it in the db. We have similar needs for user specific secrets with display author. What do you think?

Comment (patte) [reply packages/auth/lib/hashSessionId.js]: I'm not sure on this. For the record: as you pointed out in https://github.com/orbiting/republik-backend/pull/16#issuecomment-344665606 there is not much more security gained from a secret per user than a global secret in regard to stealing it. If someone steals the env she also steals the DB credentials. But attacking the secrets from the outside gets much harder. And this is true here: Even if the attacker would break her salt she could not decypher other users session id. But this can just be mitigated by a, from the sid, independent `id2` on the sessions table, that is used for the interaction with the API. This ids are only visible to their owners and actions on it are protected by meOrHasRoles, so faking a HMAC is of no use.
Similarly a user specific salt could be used, but It would not help for display author, because we show comments to not-signed-in users, which don't have a session and therefor no persistent salt. And because the display author ids need to be stable over requests and websockets also a temporary/connection salt is hard to do.
Long story short, the advantage of a user specific salt is small. What do you guys think?
NOTE: [pgcrypto](https://www.postgresql.org/docs/current/static/pgcrypto.html#idm46428634902160). A 128b key should be enough

Comment (pozylon) [reply packages/auth/lib/hashSessionId.js]: Good reads: 
- https://benlog.com/2008/06/19/dont-hash-secrets/
- https://security.stackexchange.com/questions/29951/salted-hashes-vs-hmac

From what I read we could actually use PBKDF2 which uses HMAC several times internally and use the ip as a salt which would make it extremely expensive to do bruteforcing or rainbow table attacks against the system, but the only way how we can mitigate that attack vector pretty much completely is generating a UUID in the db (id2).

An HMAC incorporates additional secret data in the form of a key. This key is combined with the message in a much deeper way than a salt, and may be used in addition to a simple salt of the message. The end result is not vulnerable to the same attack that a simple hash with a secret salt would be, and so is a much stronger method for producing an authenticated hash.

**A key is not a salt (keys are secret, salts are not)**. But the unique characteristics of HMAC make it a reasonable building block for other functions, which is what happens in PBKDF2: that's a key derivation function, commonly subverted into password hashing (a role for which it appears to be adequate). PBKDF2 includes a salt, and internally invokes HMAC (several times).

File: packages/auth/lib/hashSessionId.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+module.exports = async (id, salt) => {
+  const crypto = require('crypto')
+  return new Promise((resolve, reject) => {
+    crypto.pbkdf2(id, salt, 1000, 64, 'sha512', (err, derivedKey) => {
```
Comment (pozylon) [inline packages/auth/lib/hashSessionId.js]: - make it more expensive to calculate the hash by using pbkdf2
- don’t use a server side secret to generate the hash , as a bad guy could bruteforce the secret out of his own hash by comparing to the raw session id
- use the user’s email as salt to mitigate rainbow table attacks

@patte @tpreusse

**Comment by patte** [comment]
TODOs @pozylon 
- [ ] migrate https://github.com/orbiting/republik-backend/blob/master/modules/crowdfundings/graphql/resolvers/_mutations/updateEmail.js to use new libs
- [ ] at the moment we use `"connect-pg-simple": "^3.1.2"` check whats breaking on `4.x.x` and update if possible (I think the problem was `DATABASE_URL` not being honoured.

@TODOs patte
- [ ] DB trigger to log changes to sessions table

File: packages/auth/express/auth.js [hunk 6]
```diff
@@ -90,85 +91,42 @@ exports.configure = ({
...
+        querystring.stringify({
+          type: 'email-confirmed',
+          email,
+          context
+        })
+      )
+    } catch (e) {
+      if (e instanceof QueryEmailMismatchError) {
+        const { email } = e.meta
+        debug("session.email and query.email don't match: %O", { req: req._log(), context, ...e.meta })
```
[the remark this answers is not in the document]
Comment (tpreusse) [reply packages/auth/express/auth.js]: actually just all over the place the same. Maybe just log in lib/Sessions.js and not everywhere

File: packages/auth/graphql/resolvers/_mutations/clearSession.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    : me
+
+  try {
+    let isSessionCleared = false
+    if ((await hashSessionId(req.sessionID, me.email)) === sessionId) {
+      // current session, normal logout
+      if (await destroySession(req)) isSessionCleared = true
+    }
+    if (Roles.userIsMeOrInRoles(user, me, userAccessRoles)) {
+      if (await clearUserSession({ pgdb, store: req.sessionStore, userId: user.id || me.id, sessionId })) {
```
Comment (patte) [inline packages/auth/graphql/resolvers/_mutations/clearSession.js]: `userId: user.id || me.id` just do `userId: user.id` or am I missing something?

File: packages/auth/graphql/resolvers/_mutations/clearSessions.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  const user = foreignUserId
+    ? (await pgdb.public.users.findOne({ id: foreignUserId }))
+    : me
+
+  try {
+    let isSessionsCleared = false
+    if (me.id === user.id) {
+      // current user targeted, so we
+      // destroy the current session safely via express
+      if (await destroySession(req)) isSessionsCleared = true
```
Comment (tpreusse) [inline packages/auth/graphql/resolvers/_mutations/clearSessions.js]: wouldn't set `isSessionsCleared` here yet. Is only truly done when all are cleared

File: packages/auth/graphql/schema-types.js [hunk 9]
```diff
@@ -1,4 +1,16 @@
 module.exports = `
+
+type Session {
+  id: ID!
+  ipAddress: String!
+  userAgent: String!
+  email: String!
+  expiresAt: DateTime!
+  cookie: String
```
Comment (tpreusse) [inline packages/auth/graphql/schema-types.js]: drop

File: packages/auth/lib/Sessions.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  } catch (e) {
+    await transaction.transactionRollback()
+    throw e
+  }
+}
+
+const clearUserSession = async ({ pgdb, userId, sessionId }) => {
+  const transaction = await pgdb.transactionBegin()
+  try {
+    const existingUser = await transaction.public.users.findOne({ id: userId })
```
Comment (tpreusse) [inline packages/auth/lib/Sessions.js]: use `email = transaction.public.users.findOneFieldOnly`  http://pogi.readthedocs.io/en/latest/API/PgTable/#findonefieldonly

File: packages/auth/lib/Sessions.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    const sessions = await findAllUserSessions({ pgdb: transaction, userId })
+    const matchingSessions = sessions
+      .filter(async (session) => (
+        (await hashSessionId(session.sid, existingUser.email)) === sessionId)
+      )
+    const session = matchingSessions && matchingSessions[0]
+    if (session) await transaction.public.sessions.delete({ sid: session.sid })
```
Comment (tpreusse) [inline packages/auth/lib/Sessions.js]: shouldn't it throw if no session is found?

Comment (tpreusse) [reply packages/auth/lib/Sessions.js]: use `deleteOne` to unexpected deletes issues early

File: packages/auth/lib/hashSessionId.js (hunk 5 above)
Comment (patte) [reply packages/auth/lib/hashSessionId.js]: I'm still the opinion that it's so much more easy to trash this completely. Get rid of all the complications like async, like the recommended key length for pbkdf2 is 128bit (my email is shorter already) and just add a column to the sessions table called `id uuid not null`. This id is just used for referencing the session via the API, and public. It has nothing to do with the session id, that is just db.sid <-> cookie.

File: packages/auth/express/auth.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (pozylon) [reply packages/auth/express/auth.js]: I think it's important to log at each endpoint so we know explicitly which errors can be thrown and I don't have to pass req around. logging at the endpoint allows us to additional log context specific stuff and allows for better tracing.

[the remark this answers is not in the document]
Comment (tpreusse) [reply packages/auth/express/auth.js]: ok with me. Was initially confused thought it double logged.

File: packages/auth/lib/Sessions.js (hunk 11 above)
Comment (pozylon) [reply packages/auth/lib/Sessions.js]: 💯
